assets/model_monitoring/components/src/action_analyzer/action_detector_component/run.py

```diff
@@ -0,0 +1,166 @@
+import pandas
+from typing import List
+from action_analyzer.contracts.action_detectors.low_retrieval_score_index_action_detector import (
+    LowRetrievalScoreIndexActionDetector
+)
+from action_analyzer.contracts.llm_client import LLMClient
+from action_analyzer.contracts.utils.detector_utils import (
+    get_index_id_from_index_content,
+    deduplicate_actions,
+    write_actions,
+    add_action_tag_to_run
+)
+from shared_utilities.constants import (
+    RETRIEVAL_SPAN_TYPE,
+    ROOT_SPAN_COLUMN,
+    GSQ_METRICS_LIST,
+    INDEX_ID_COLUMN
+)
+from shared_utilities.store_url import StoreUrl
+from shared_utilities.io_utils import try_read_mltable_in_spark
+from shared_utilities.span_tree_utils import SpanTree
+
+
+def parse_index_id(root_span: str) -> List[str]:
+    """Parse the span tree to get index id.
+
+    Args:
+        root_span(str): the span tree in json string format.
+
+    Returns:
+        list(str): list of parsed indexes in the span tree.
+    """
+    try:
+        tree = SpanTree.create_tree_from_json_string(root_span)
+        index_list = []
+        for span in tree:
+            if span.span_type == RETRIEVAL_SPAN_TYPE:
+                parent_id = span.parent_id
+                if not parent_id:
+                    print("No look up span found, skip action analyzer.")
+                    return None
+                index_span = tree.get_span_tree_node_by_span_id(parent_id)
+                print(index_span)
+                index_input = json.loads(index_span.input)
+                print(index_input)
+                index_content = index_input['mlindex_content']
+                index_list.append(get_index_id_from_index_content(index_content))
+        return index_list
+    except KeyError as e:
+        print("Required field not found: ", e)
+        return None
+
+
+def get_unique_indexes(df: pandas.DataFrame) -> List[str]:
+    """Parse the span tree to find all unique indexes.
+
+    Args:
+        df(pandas.DataFrame): input pandas dataframe in trace level.
+
+    Returns:
+        List[str]: list of unique indexes.
+    """
+    df[INDEX_ID_COLUMN] = df[ROOT_SPAN_COLUMN].apply(parse_index_id)
+    # expand each trace row to index level
+    df = df.explode(INDEX_ID_COLUMN)
+    return df[INDEX_ID_COLUMN].unique().tolist()
+
+
+def get_violated_metrics(signal_out_url: str, signal_name: str) -> List[str]:
+    """Get the violated metrics names from the gsq output.
+
+    Args:
+        signal_out_url(str): gsq output url.
+        signal_name(str): signal name defined by user.
+    Returns:
+        List[str]: list of violated metrics.
+    """
+    violated_metrics = []
+    try:
+        store_url = StoreUrl(signal_out_url)
+        gsq_output = store_url.read_file_content(f"{signal_name}.json")
+        gsq_output_json = json.loads(gsq_output)
+        metrics_dict = gsq_output_json["metrics"]
+        for metrics in GSQ_METRICS_LIST:
+            pass_rate_metrics = f"Aggregated{metrics}PassRate"
```

Comment:
They changed the passrate to average score

Reply:
changed to average score and update UT